File: frontend/src/api/github.ts
import {
  GitHubUserResponse,
  GitHubRepoResponse,
  GitHubUser,
  GitHubRepo,
  LanguageStat,
} from '../types/github';

const GITHUB_API_BASE = 'https://api.github.com';
const CACHE_TTL_MS = 15 * 60 * 1000; // 15 minutes cache TTL

// Map common languages to hex colors
const LANGUAGE_COLORS: Record<string, string> = {
  TypeScript: '#3178c6',
  JavaScript: '#f1e05a',
  Python: '#3572A5',
  HTML: '#e34c26',
  CSS: '#563d7c',
  Shell: '#89e051',
  Go: '#00ADD8',
  Rust: '#dea584',
  C: '#555555',
  'C++': '#f34b7d',
  Java: '#b07219',
  Ruby: '#701516',
};

// Check if a date string falls within the last 30 days
export function isWithinPast30Days(dateString: string): boolean {
  if (!dateString) return false;
  const date = new Date(dateString);
  const thirtyDaysAgo = new Date();
  thirtyDaysAgo.setDate(thirtyDaysAgo.getDate() - 30);
  return date >= thirtyDaysAgo;
}

// Format date into human-readable relative string
export function formatRelativeTime(dateString: string): string {
  if (!dateString) return 'recently';
  const date = new Date(dateString);
  const now = new Date();
  const diffInSeconds = Math.floor((now.getTime() - date.getTime()) / 1000);

  if (diffInSeconds < 60) return 'just now';
  if (diffInSeconds < 3600) return `${Math.floor(diffInSeconds / 60)}m ago`;
  if (diffInSeconds < 86400) return `${Math.floor(diffInSeconds / 3600)}h ago`;
  if (diffInSeconds < 2592000) return `${Math.floor(diffInSeconds / 86400)}d ago`;
  if (diffInSeconds < 31536000) return `${Math.floor(diffInSeconds / 2592000)}mo ago`;
  return `${Math.floor(diffInSeconds / 31536000)}y ago`;
}

interface CacheItem<T> {
  data: T;
  timestamp: number;
}

function getCache<T>(key: string): T | null {
  try {
    const raw = sessionStorage.getItem(key);
    if (!raw) return null;
    const item: CacheItem<T> = JSON.parse(raw);
    if (Date.now() - item.timestamp > CACHE_TTL_MS) {
      sessionStorage.removeItem(key);
      return null;
    }
    return item.data;
  } catch {
    return null;
  }
}

function setCache<T>(key: string, data: T): void {
  try {
    const item: CacheItem<T> = { data, timestamp: Date.now() };
    sessionStorage.setItem(key, JSON.stringify(item));
  } catch {
    // Ignore storage quota errors
  }
}

// Compute language stats from repositories
function computeLanguageStats(repos: GitHubRepoResponse[]): LanguageStat[] {
  const counts: Record<string, number> = {};
  let total = 0;

  repos.forEach((repo) => {
    if (repo.language && !repo.fork) {
      counts[repo.language] = (counts[repo.language] || 0) + 1;
      total += 1;
    }
  });

  if (total === 0) return [];

  const sorted = Object.entries(counts)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 5);

  return sorted.map(([lang, count]) => ({
    language: lang,
    count,
    percentage: Math.round((count / total) * 100),
    color: LANGUAGE_COLORS[lang] || '#8b949e',
  }));
}

// Transform raw GitHub repo response to internal GitHubRepo model
export function transformGitHubRepo(raw: GitHubRepoResponse): GitHubRepo {
  const isRecentlyUpdated =
    isWithinPast30Days(raw.pushed_at) || isWithinPast30Days(raw.updated_at);

  return {
    id: raw.id,
    name: raw.name,
    fullName: raw.full_name,
    description: raw.description || 'No description provided.',
    githubUrl: raw.html_url,
    demoUrl: raw.homepage && raw.homepage.trim().length > 0 ? raw.homepage : null,
    stars: raw.stargazers_count,
    forks: raw.forks_count,
    primaryLanguage: raw.language || 'Markdown',
    topics: raw.topics || [],
    isFork: raw.fork,
    updatedAt: raw.updated_at,
    pushedAt: raw.pushed_at,
    formattedLastUpdated: formatRelativeTime(raw.pushed_at || raw.updated_at),
    isRecentlyUpdated,
  };
}

// Fetch GitHub profile info for a given username
export async function fetchGitHubUser(username: string): Promise<GitHubUser> {
  const cleanUsername = username.trim().toLowerCase();
  const cacheKey = `gh_user_${cleanUsername}`;
  const cached = getCache<GitHubUser>(cacheKey);
  if (cached) return cached;

  const res = await fetch(`${GITHUB_API_BASE}/users/${cleanUsername}`);
  if (!res.ok) {
    if (res.status === 404) {
      throw new Error(`GitHub user "${username}" was not found.`);
    }
    if (res.status === 403) {
      throw new Error('GitHub API rate limit exceeded. Please try again in a few minutes.');
    }
    throw new Error(`Failed to load GitHub user (HTTP ${res.status}).`);
  }

  const rawUser: GitHubUserResponse = await res.json();
  const reposResponse = await fetch(`${GITHUB_API_BASE}/users/${cleanUsername}/repos?per_page=100&sort=updated`);
  const rawRepos: GitHubRepoResponse[] = reposResponse.ok ? await reposResponse.json() : [];

  const userModel: GitHubUser = {
    username: rawUser.login,
    displayName: rawUser.name || rawUser.login,
    avatarUrl: rawUser.avatar_url,
    profileUrl: rawUser.html_url,
    bio: rawUser.bio || 'Software developer & technology enthusiast.',
    publicRepos: rawUser.public_repos,
    followers: rawUser.followers,
    following: rawUser.following,
    location: rawUser.location || undefined,
    blogUrl: rawUser.blog || undefined,
    topLanguages: computeLanguageStats(rawRepos),
  };

  setCache(cacheKey, userModel);
  return userModel;
}

// Fetch public repositories for a given username
export async function fetchGitHubRepos(username: string): Promise<GitHubRepo[]> {
  const cleanUsername = username.trim().toLowerCase();
  const cacheKey = `gh_repos_${cleanUsername}`;
  const cached = getCache<GitHubRepo[]>(cacheKey);
  if (cached) return cached;

  const res = await fetch(
    `${GITHUB_API_BASE}/users/${cleanUsername}/repos?per_page=100&sort=pushed`
  );

  if (!res.ok) {
    if (res.status === 404) {
      throw new Error(`GitHub repositories for "${username}" were not found.`);
    }
    if (res.status === 403) {
      throw new Error('GitHub API rate limit exceeded. Please try again in a few minutes.');
    }
    throw new Error(`Failed to load GitHub repositories (HTTP ${res.status}).`);
  }

  const rawRepos: GitHubRepoResponse[] = await res.json();
  const repos = rawRepos
    .filter((r) => !r.fork) // Exclude forks by default for cleaner portfolio view
    .map(transformGitHubRepo);

  setCache(cacheKey, repos);
  return repos;
}

File: backend/api/endpoints/github.py
"""GitHub proxy endpoint.

Proxies GitHub REST API requests server-side so the frontend avoids the
unauthenticated 60 req/hr IP-shared limit. When ``GITHUB_TOKEN`` is set, the
authenticated 5000 req/hr budget is used. Results are cached in-memory for
15 minutes.
"""

from __future__ import annotations

import logging
import time
from datetime import datetime, timezone, timedelta

import httpx
from fastapi import APIRouter, HTTPException, Query
from starlette.requests import Request

from config import settings
from core.rate_limit import limiter
from schemas.github import (
    GitHubRepoSummary,
    GitHubSummaryResponse,
    GitHubUserSummary,
    LanguageStat,
)

router = APIRouter()

logger = logging.getLogger(__name__)

GITHUB_API_BASE = "https://api.github.com"
CACHE_TTL_SECONDS = 15 * 60  # 15 minutes

# Language -> hex color (mirrors the frontend color map).
LANGUAGE_COLORS: dict[str, str] = {
    "TypeScript": "#3178c6",
    "JavaScript": "#f1e05a",
    "Python": "#3572A5",
    "HTML": "#e34c26",
    "CSS": "#563d7c",
    "Shell": "#89e051",
    "Go": "#00ADD8",
    "Rust": "#dea584",
    "C": "#555555",
    "C++": "#f34b7d",
    "Java": "#b07219",
    "Ruby": "#701516",
}

# In-memory cache: { username: { "data": ..., "expires": epoch } }.
_cache: dict[str, dict] = {"hits": 0, "misses": 0, "entries": {}}


def _github_headers() -> dict[str, str]:
    headers = {"Accept": "application/vnd.github+json", "User-Agent": "personal-os-backend"}
    if settings.GITHUB_TOKEN:
        headers["Authorization"] = f"Bearer {settings.GITHUB_TOKEN}"
    return headers


def _is_within_past_30_days(date_string: str) -> bool:
    if not date_string:
        return False
    try:
        date = datetime.fromisoformat(date_string.replace("Z", "+00:00"))
    except (ValueError, TypeError):
        return False
    cutoff = datetime.now(timezone.utc) - timedelta(days=30)
    return date >= cutoff


def _format_relative_time(date_string: str) -> str:
    if not date_string:
        return "recently"
    try:
        date = datetime.fromisoformat(date_string.replace("Z", "+00:00"))
    except (ValueError, TypeError):
        return "recently"
    now = datetime.now(timezone.utc)
    diff = (now - date).total_seconds()
    if diff < 60:
        return "just now"
    if diff < 3600:
        return f"{int(diff // 60)}m ago"
    if diff < 86400:
        return f"{int(diff // 3600)}h ago"
    if diff < 2592000:
        return f"{int(diff // 86400)}d ago"
    if diff < 31536000:
        return f"{int(diff // 2592000)}mo ago"
    return f"{int(diff // 31536000)}y ago"


def _compute_language_stats(repos: list[dict]) -> list[LanguageStat]:
    counts: dict[str, int] = {}
    total = 0
    for repo in repos:
        lang = repo.get("language")
        if lang and not repo.get("fork"):
            counts[lang] = counts.get(lang, 0) + 1
            total += 1
    if total == 0:
        return []
    sorted_langs = sorted(counts.items(), key=lambda x: x[1], reverse=True)[:5]
    return [
        LanguageStat(
            language=lang,
            count=count,
            percentage=round((count / total) * 100),
            color=LANGUAGE_COLORS.get(lang, "#8b949e"),
        )
        for lang, count in sorted_langs
    ]


def _transform_repo(raw: dict) -> GitHubRepoSummary:
    pushed_at = raw.get("pushed_at", "")
    updated_at = raw.get("updated_at", "")
    recently_updated = _is_within_past_30_days(pushed_at) or _is_within_past_30_days(updated_at)
    homepage = raw.get("homepage") or ""
    return GitHubRepoSummary(
        id=raw["id"],
        name=raw["name"],
        fullName=raw["full_name"],
        description=raw.get("description") or "No description provided.",
        githubUrl=raw["html_url"],
        demoUrl=homepage.strip() if homepage.strip() else None,
        stars=raw.get("stargazers_count", 0),
        forks=raw.get("forks_count", 0),
        primaryLanguage=raw.get("language") or "Markdown",
        topics=raw.get("topics") or [],
        isFork=raw.get("fork", False),
        updatedAt=updated_at,
        pushedAt=pushed_at,
        formattedLastUpdated=_format_relative_time(pushed_at or updated_at),
        isRecentlyUpdated=recently_updated,
    )


async def _fetch_github_summary(username: str) -> GitHubSummaryResponse:
    """Fetch user + repos from GitHub and transform into view models."""
    headers = _github_headers()
    async with httpx.AsyncClient(timeout=10.0) as client:
        user_resp = await client.get(
            f"{GITHUB_API_BASE}/users/{username}", headers=headers
        )
        if user_resp.status_code == 404:
            raise HTTPException(status_code=404, detail=f'GitHub user "{username}" was not found.')
        if user_resp.status_code == 403:
            raise HTTPException(
                status_code=403,
                detail="GitHub API rate limit exceeded. Please try again in a few minutes.",
            )
        if user_resp.status_code != 200:
            raise HTTPException(
                status_code=502,
                detail=f"GitHub user request failed (HTTP {user_resp.status_code}).",
            )
        raw_user = user_resp.json()

        repos_resp = await client.get(
            f"{GITHUB_API_BASE}/users/{username}/repos",
            params={"per_page": 100, "sort": "pushed"},
            headers=headers,
        )
        if repos_resp.status_code == 403:
            raise HTTPException(
                status_code=403,
                detail="GitHub API rate limit exceeded. Please try again in a few minutes.",
            )
        if repos_resp.status_code != 200:
            raise HTTPException(
                status_code=502,
                detail=f"GitHub repos request failed (HTTP {repos_resp.status_code}).",
            )
        raw_repos: list[dict] = repos_resp.json()

    user_model = GitHubUserSummary(
        username=raw_user["login"],
        displayName=raw_user.get("name") or raw_user["login"],
        avatarUrl=raw_user["avatar_url"],
        profileUrl=raw_user["html_url"],
        bio=raw_user.get("bio") or "Software developer & technology enthusiast.",
        publicRepos=raw_user.get("public_repos", 0),
        followers=raw_user.get("followers", 0),
        following=raw_user.get("following", 0),
        location=raw_user.get("location") or None,
        blogUrl=raw_user.get("blog") or None,
        topLanguages=_compute_language_stats(raw_repos),
    )
    repos = [_transform_repo(r) for r in raw_repos if not r.get("fork")]

    return GitHubSummaryResponse(user=user_model, repos=repos)


@router.get(
    "/github-summary",
    response_model=GitHubSummaryResponse,
    summary="Proxy GitHub user + repos summary (avoids client-side rate limits)",
)
@limiter.limit("30/minute")
async def get_github_summary(
    request: Request,
    username: str = Query(..., min_length=1, max_length=100, pattern=r"^[\w\-]+$"),
):
    # Check in-memory cache first.
    clean = username.strip().lower()
    entries: dict = _cache["entries"]
    cached = entries.get(clean)
    if cached and time.time() < cached["expires"]:
        _cache["hits"] += 1
        result = cached["data"]
        result.cached = True
        result.stale = False
        return result

    _cache["misses"] += 1
    try:
        result = await _fetch_github_summary(clean)
    except (HTTPException, httpx.HTTPError) as exc:
        # Stale-while-error: when the GitHub upstream is unavailable (rate
        # limit / 5xx / network error) serve the expired cache entry rather
        # than hard-failing clients. A 404 means the user genuinely does not
        # exist, so it is never masked by stale data.
        status_code = exc.status_code if isinstance(exc, HTTPException) else 502
        if cached and status_code != 404:
            logger.warning(
                "GitHub upstream failed (HTTP %s) — serving stale cache for %s",
                status_code,
                clean,
            )
            result = cached["data"]
            result.cached = True
            result.stale = True
            return result
        if not isinstance(exc, HTTPException):
            raise HTTPException(
                status_code=502, detail="GitHub upstream request failed."
            ) from exc
        raise

    entries[clean] = {"data": result, "expires": time.time() + CACHE_TTL_SECONDS}
    result.cached = False
    result.stale = False
    return result


def get_cache_stats() -> tuple[int, int]:
    """Return (hits, misses) for telemetry reporting."""
    return _cache["hits"], _cache["misses"]


def clear_github_cache() -> None:
    """Clear the in-memory GitHub cache (for testing)."""
    entries: dict = _cache["entries"]
    entries.clear()
    _cache["hits"] = 0
    _cache["misses"] = 0
